README.md
# Weightify: The Ultimate Weight Conversion Tool

[![Latest Version on Packagist](https://img.shields.io/packagist/v/karimtarekdev/weightify.svg?style=flat-square)](https://packagist.org/packages/karimtarekdev/weightify)
[![Tests](https://img.shields.io/github/actions/workflow/status/karimtarekdev/weightify/run-tests.yml?branch=main&label=tests&style=flat-square)](https://github.com/karimtarekdev/weightify/actions/workflows/run-tests.yml)
[![Total Downloads](https://img.shields.io/packagist/dt/karimtarekdev/weightify.svg?style=flat-square)](https://packagist.org/packages/karimtarekdev/weightify)

<p align="center">
    <img src="https://i.imgur.com/NmcsUTp.png">
</p>

"Weightify: The Ultimate Weight Conversion Tool" is a streamlined and efficient PHP package for accurate weight conversions. It supports a wide array of units such as Metric tons, Kilograms, Grams, Milligrams, Micrograms, Stones, Pounds, and Ounces, making it an ideal solution for various conversion needs in a user-friendly format.

## Features
- Broad Range of Units: Supports conversions between Metric tons, Kilograms, Grams, Milligrams, Micrograms, Stones, Pounds, and Ounces.
- High Precision: Accurate conversions even for very large (metric tons) or very small (micrograms) units.
- User-Friendly: Simple and straightforward interface for easy integration into any PHP project.
- Exception Handling: Robust error management for unsupported units.

## Installation

You can install the package via composer:

```bash
composer require karimtarekdev/weightify
```

## Supported Units
- Metric Tons (metric_ton)
- Kilograms (kg)
- Grams (gram)
- Milligrams (mg)
- Micrograms (microgram)
- Stones (stone)
- Pounds (pound)
- Ounces (ounce)

## Basic Conversion
To perform a conversion, use the convert method with the desired units. Here are some examples showcasing the use of different units:

```php
$converter = new KarimTarekDev\Weightify\Weightify();

// Convert 2 Metric tons to Pounds
echo $converter->convert(2, 'metric_ton', 'pound') . " lbs\n";

// Convert 1500 Milligrams to Grams
echo $converter->convert(1500, 'mg', 'gram') . " grams\n";

// Convert 3 Stones to Kilograms
echo $converter->convert(3, 'stone', 'kg') . " kg\n";

// Convert 200 Ounces to Micrograms
echo $converter->convert(200, 'ounce', 'microgram') . " micrograms\n";
```
By providing the value, source unit, and target unit, you can perform a wide range of conversions with "Weightify."


## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

- [Karim Tarek](https://github.com/KarimTarekDev)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `Weightify` will be documented in this file.

## 1.0.0 - 2023-11-20

- Initial release
 


tests/ExampleTest.php
<?php

namespace KarimTarekDev\Weightify\Tests;

use PHPUnit\Framework\TestCase;

class ExampleTest extends TestCase
{
}


tests/WeightifyTest.php
<?php

namespace KarimTarekDev\Weightify\Tests;

use Exception;
use KarimTarekDev\Weightify\Weightify;
use PHPUnit\Framework\TestCase;

class WeightifyTest extends TestCase
{
    private Weightify $converter;

    protected function setUp(): void
    {
        $this->converter = new Weightify();
    }

    public static function conversionDataProvider(): array
    {
        return [
            ['kg', 'pound', 1, 2.20462],
            ['pound', 'kg', 2.20462, 1],
            ['gram', 'mg', 1, 1000],
            ['mg', 'microgram', 1, 1000],
            ['stone', 'pound', 1, 14],
            ['ounce', 'pound', 16, 1],
            ['metric_ton', 'kg', 1, 1000],
        ];
    }

    /**
     * @dataProvider conversionDataProvider
     * @throws Exception
     */
    public function testUnitConversions($fromUnit, $toUnit, $input, $expected)
    {
        $result = $this->converter->convert($input, $fromUnit, $toUnit);
        $this->assertEqualsWithDelta($expected, $result, 0.001, '');
    }

    public function testUnsupportedUnit()
    {
        $this->expectException(Exception::class);
        $this->converter->convert(1, 'unknown_unit', 'kg');
    }
}


src/Weightify.php
<?php

namespace KarimTarekDev\Weightify;

use Exception;

class Weightify
{
    private array $units = ['metric_ton', 'kg', 'gram', 'mg', 'microgram', 'stone', 'pound', 'ounce'];

    private function convertToKg($value, $unit)
    {
        switch ($unit) {
            case 'metric_ton':
                return $value * 1000;    // 1 metric ton = 1000 kg
            case 'gram':
                return $value * 0.001;   // 1 gram = 0.001 kg
            case 'mg':
                return $value * 1e-6;    // 1 milligram = 1e-6 kg
            case 'microgram':
                return $value * 1e-9;    // 1 microgram = 1e-9 kg
            case 'stone':
                return $value * 6.35;    // 1 stone = 6.35 kg
            case 'pound':
                return $value * 0.453592;// 1 pound = 0.453592 kg
            case 'ounce':
                return $value * 0.0283495;// 1 ounce = 0.0283495 kg
            case 'kg':
            default:
                return $value;
        }
    }

    private function convertFromKg($value, $unit)
    {
        switch ($unit) {
            case 'metric_ton':
                return $value / 1000;    // 1 kg = 1/1000 metric ton
            case 'gram':
                return $value / 0.001;   // 1 kg = 1000 grams
            case 'mg':
                return $value / 1e-6;    // 1 kg = 1e6 milligrams
            case 'microgram':
                return $value / 1e-9;    // 1 kg = 1e9 micrograms
            case 'stone':
                return $value / 6.35;    // 1 kg = 1/6.35 stones
            case 'pound':
                return $value / 0.453592;// 1 kg = 1/0.453592 pounds
            case 'ounce':
                return $value / 0.0283495;// 1 kg = 1/0.0283495 ounces
            case 'kg':
            default:
                return $value;
        }
    }

    /**
     * @throws Exception
     */
    public function convert($value, $fromUnit, $toUnit)
    {
        if (! in_array($fromUnit, $this->units) || ! in_array($toUnit, $this->units)) {
            throw new Exception("Unsupported unit.");
        }

        $valueInKg = $this->convertToKg($value, $fromUnit);

        return $this->convertFromKg($valueInKg, $toUnit);
    }
}
